pkg/protocol/protocol_attestations.go

```diff
@@ -0,0 +1,260 @@
+// responses.
+type AttestationsProtocol struct {
+	// protocol contains a reference to the Protocol instance that this component belongs to.
+	protocol *Protocol
+
+	// workerPool contains the worker pool that is used to process attestation requests and responses asynchronously.
+	workerPool *workerpool.WorkerPool
+
+	// ticker contains the ticker that is used to send attestation requests.
+	ticker *eventticker.EventTicker[iotago.SlotIndex, iotago.CommitmentID]
+
+	// commitmentVerifiers contains the commitment verifiers that are used to verify received attestations.
+	commitmentVerifiers *shrinkingmap.ShrinkingMap[iotago.CommitmentID, *CommitmentVerifier]
+
+	// Logger embeds a logger that can be used to log messages emitted by this chain.
+	log.Logger
+}
+
+// newAttestationsProtocol creates a new attestation protocol instance for the given protocol.
+func newAttestationsProtocol(protocol *Protocol) *AttestationsProtocol {
+	a := &AttestationsProtocol{
+		Logger:              lo.Return1(protocol.Logger.NewChildLogger("Attestations")),
+		protocol:            protocol,
+		workerPool:          protocol.Workers.CreatePool("Attestations"),
+		ticker:              eventticker.New[iotago.SlotIndex, iotago.CommitmentID](),
+		commitmentVerifiers: shrinkingmap.New[iotago.CommitmentID, *CommitmentVerifier](),
+	}
+
+	a.ticker.Events.Tick.Hook(a.sendRequest)
+
+	protocol.Constructed.OnTrigger(func() {
+		protocol.Chains.WithElements(func(chain *Chain) (shutdown func()) {
+			return chain.RequestAttestations.WithNonEmptyValue(func(requestAttestations bool) (shutdown func()) {
+				return a.setupCommitmentVerifier(chain)
+			})
+		})
+
+		protocol.Commitments.WithElements(func(commitment *Commitment) (shutdown func()) {
+			return commitment.RequestAttestations.OnUpdate(func(_ bool, requestAttestations bool) {
+				if requestAttestations {
+					if commitment.CumulativeWeight() == 0 {
+						commitment.IsAttested.Set(true)
+					} else {
+						a.ticker.StartTicker(commitment.ID())
+					}
+				} else {
+					a.ticker.StopTicker(commitment.ID())
+				}
+			})
+		})
+	})
+
+	return a
+}
+
+// ProcessResponse processes the given attestation response.
+func (a *AttestationsProtocol) ProcessResponse(commitmentModel *model.Commitment, attestations []*iotago.Attestation, merkleProof *merklehasher.Proof[iotago.Identifier], from peer.ID) {
+	a.workerPool.Submit(func() {
+		commitment, _, err := a.protocol.Commitments.publishCommitmentModel(commitmentModel)
+		if err != nil {
+			a.LogDebug("failed to publish commitment when processing attestations", "commitmentID", commitmentModel.ID(), "peer", from, "error", err)
+
+			return
+		}
+
+		if commitment.AttestedWeight.Compute(func(currentWeight uint64) uint64 {
+			if !commitment.RequestAttestations.Get() {
+				a.LogTrace("received attestations for previously attested commitment", "commitment", commitment.LogName())
+
+				return currentWeight
```

Comment:
The original idea was to allow attestations to be "updated" if we for example receive more attestations from a different neighbor with a higher weight and for this, access to the weight needs to be threat safe + we want to keep the current value in the error case.

Since I currently just ported the existing logic without implementing such an additional update mechanism to keep the logic changes minimal, we could also synchronize over the attested flag but since that change is already planned, I would keep it as is.